pkg/ruler/ruler.go

```diff
@@ -518,6 +556,68 @@ func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRu
 	return ownsRuleGroup, nil
 }
 
+func (r *Ruler) LivenessCheck(_ context.Context, request *LivenessCheckRequest) (*LivenessCheckResponse, error) {
+	if r.lifecycler.ServiceContext().Err() != nil || r.subservices.IsStopped() {
+		return nil, errors.New("ruler's context is canceled and might be stopping soon")
+	}
+	if !r.subservices.IsHealthy() {
+		return nil, errors.New("not all subservices are in healthy state")
+	}
+	return &LivenessCheckResponse{State: int32(r.State())}, nil
+}
+
+// This function performs a liveness check against the provided replicas. If any one of the replicas responds with a state = Running, then
+// this Ruler should not take ownership of the rule group. Otherwise, this Ruler must take ownership of the rule group to avoid missing evaluations
+func (r *Ruler) nonPrimaryInstanceOwnsRuleGroup(g *rulespb.RuleGroupDesc, replicas []string, selfAddress string) bool {
+	userID := g.User
+
+	jobs := concurrency.CreateJobsFromStrings(replicas)
+
+	errorChan := make(chan error, len(jobs))
+	responseChan := make(chan *LivenessCheckResponse, len(jobs))
+
+	ctx := user.InjectOrgID(context.Background(), userID)
+	ctx, cancel := context.WithTimeout(ctx, livenessCheckTimeout)
+	defer cancel()
+
+	err := concurrency.ForEach(ctx, jobs, len(jobs), func(ctx context.Context, job interface{}) error {
+		addr := job.(string)
+		rulerClient, err := r.GetClientFor(addr)
+		if err != nil {
+			errorChan <- err
+			level.Debug(r.Logger()).Log("msg", "unable to get client for ruler", "ruler addr", addr)
+			return nil
+		}
+		level.Debug(r.Logger()).Log("msg", "performing liveness check against", "addr", addr, "for", g.Name, "instance addr", selfAddress)
+
+		resp, err := rulerClient.LivenessCheck(ctx, &LivenessCheckRequest{
+			RulerAddress: selfAddress,
```

Comment:
any reason why we are passing the address of the current ruler as parameter here? I checked and this is not being used.

Reply:
It's not being used now but I was using it for troubleshooting/logging on the server side. It could still be useful for troubleshooting

Reply:
If it is supposed to be used for troubleshooting we can add a debug log statement in the handler no? I think leaving in this state makes me think of YAGNI

Reply:
I also thing we can remove for now and add when needed.